### PR TITLE
restore to normal tokenlization when validation

### PR DESCRIPTION
**Pull request by znculee**
When the [BPE](https://github.com/rsennrich/subword-nmt) is used to preprocess the corpus, our model only can be validated on the dataset processed by BPE, which is not the same as the real testing, namely testing BLEU after restoring BPE to normal tokenlization. Besides, in the experiments, I found using BPE restored validation may be better.

In this pull request, I add a new param called `restore_bpe`. When it is set to be `True`, `decoded_symbols`  in `EvaluationHook` will be restored to the normal tokenlization. And the default value of `restore_bpe` is false, which would not affect the original implementation.

### CODE REVIEW

**Comment by Glaceon31** [comment]
Thanks for your contribution! This is very useful as BPE has become a standard preprocessing step in NMT. We will merge it after testing.

**Comment by Glaceon31** [comment]
Sorry for the late reply. I have tested the code but find some problems.

I train on a toy dataset with your code:

===train.zh===
我 也 喜欢 音乐
他 是 一个 老师

===train.en===
i also l@@ i@@ k@@ e m@@ u@@ s@@ i@@ c
he is a teacher

===valid.zh===(copy of train.zh)
我 也 喜欢 音乐
他 是 一个 老师

===valid.en===(adding a period to each sentence)
i also l@@ i@@ k@@ e m@@ u@@ s@@ i@@ c .
he is a teacher .

The command is 
`python thumt/bin/trainer.py --input train.zh train.en --output model_transformer --vocabulary vocab.zh.txt vocab.en.txt --model transformer --validation valid.zh --references valid.en --parameters constant_batch_size=false,batch_size=50,max_length=100,update_cycle=1,device_list=[0],train_steps=1000,keep_top_checkpoint_max=100,eval_steps=100,restore_bpe=True`

The model learns to overfit on the training set after one hundred iterations. The BLEU score on validation is 0.8144 before restoring BPE and 0.5454 after restoring BPE (calculated with multi-bleu.perl). However, the validation score output in the training log is 0.000000. When I turn off `restore_bpe`, the validation score is 0.814410, which is same as the BLEU score calculated with multi-bleu.perl.

Please point out if I make some mistakes in the test.

**Comment by znculee** [comment]
@Glaceon31 Thanks for your testing. However, I have tested your simple example, and there should not be any problem. In addition, I also have well used this in large scale experiments. Where did you check the validation score? I think you may check the `eval_history` file, and where the best bleu should be equal to the bleu which is translated by `translator.py`.

The key code of restoring BPE is `decoded_symbols[i] = ' '.join(l).replace('@@ ', '').split()`, which restore BPE by replacing @@<space> to nothing.

**Comment by Glaceon31** [comment]
I understand the key code of restoring BPE (line 171~173 in hook.py). I check the validation score in `model/eval/log` and `model/eval/record`, the content is `2018-08-29 16:48:14.136661: BLEU at step 400: 0.000000` and `"model.ckpt-400": 0.000000` respectively. I also check the validation score in the training output and find `INFO:tensorflow:BLEU at step 200: 0.000000`. I clone the code from the `restore_bpe` branch on your repository for testing.

**Comment by minicheshire** [comment]
Thanks @znculee ! After more tests, we've found that you're correct... We made a mistake before by treating the target side of validation set **with BPE operations** as the ground-truth during evaluation, and had been comparing the translation output (which has been restored to normal tokenization!) with it...

Now it's been fixed. I've added the key line in your PR. See the recent commit: [d0c7ea6](https://github.com/THUNLP-MT/THUMT/commit/d0c7ea6c2f2f15fe6800f0190cbb6b07f34b0228); we've also updated the user manual about this. Thanks again for your contribution!

BTW, there have been more new features added to THUMT; we'll update the user manual in the near future. Please stay tuned; suggestions are welcome!
